core/jvm/test/PoolingTest.kt

```diff
@@ -40,4 +45,107 @@ class PoolingTest {
         otherBuffer.clear()
         assertTrue(poolSize < SegmentPool.byteCount)
     }
+
+    @Test
+    fun segmentAcquisitionAndRelease() {
+        val secondTierSize = SegmentPool.SECOND_LEVEL_POOL_TOTAL_SIZE
+        val firstTierSize = SegmentPool.MAX_SIZE
+
+        // fill the pool by requiring max possible segments count and then
+        // releasing them all
+        val segments = mutableSetOf<Segment>()
+        var size = 0
+        while (size < secondTierSize + firstTierSize) {
+            val segment = SegmentPool.take()
+            size += segment.dataAsByteArray(false).size
+            segments.add(segment)
+        }
+        segments.forEach(SegmentPool::recycle)
+
+
+        // take the same number of segments again and check that nothing new was allocated
+        val segments2 = mutableSetOf<Segment>()
+        size = 0
+        while (size < secondTierSize + firstTierSize) {
+            val segment = SegmentPool.take()
+            size += segment.remainingCapacity
+            segments2.add(segment)
+        }
+        segments2.forEach(SegmentPool::recycle)
+
+        assertEquals(segments, segments2)
+    }
+
+    @Test
+    fun contendedUseTest() {
+        val threadsCount = Runtime.getRuntime().availableProcessors()
+        val segmentsPerThread = SegmentPool.SECOND_LEVEL_POOL_TOTAL_SIZE / Segment.SIZE / threadsCount
+        val observedSegments = ConcurrentHashMap<Segment, Any>()
+
+        val pool = Executors.newFixedThreadPool(threadsCount)
+        repeat(threadsCount) {
+            pool.submit {
+                repeat(10000) {
+                    val segments = mutableListOf<Segment>()
+                    repeat(segmentsPerThread) {
+                        val seg = SegmentPool.take()
+                        segments.add(seg)
+                        observedSegments[seg] = seg
+                    }
+                    segments.forEach { SegmentPool.recycle(it) }
+                }
+            }
+        }
+        pool.shutdown()
+        pool.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS)
+
+        val maxPoolSize =
```

Comment:
Should we try to limit the number of segments more precisely?
E.g. use `(SegmentPool.SECOND_LEVEL_POOL_TOTAL_SIZE + SegmentPool.MAX_SIZE * threadsCount) / Segment.SIZE`

Reply:
Instead, I increased the number of threads to make sure that it'll exceed the number of buckets